From 0015 Mon Sep 17 00:00:00 2001
From: blaze <blaze@example.com>
Date: Mon, 27 Jul 2026 02:26:00 +0800
Subject: [PATCH] blaze: update arrow dependency to vendored v55.2.0-blaze

---
 Cargo.toml | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/Cargo.toml b/Cargo.toml
index 11cd3c637..b2fbf0df3 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -230,3 +230,19 @@ unexpected_cfgs = { level = "warn", check-cfg = [
     "cfg(tarpaulin_include)",
 ] }
 unused_qualifications = "deny"
+
+[patch.crates-io]
+# arrow: vendored from v55.2.0-blaze
+arrow = { path = "../arrow-rs/arrow" }
+arrow-arith = { path = "../arrow-rs/arrow-arith" }
+arrow-array = { path = "../arrow-rs/arrow-array" }
+arrow-buffer = { path = "../arrow-rs/arrow-buffer" }
+arrow-cast = { path = "../arrow-rs/arrow-cast" }
+arrow-data = { path = "../arrow-rs/arrow-data" }
+arrow-ipc = { path = "../arrow-rs/arrow-ipc" }
+arrow-ord = { path = "../arrow-rs/arrow-ord" }
+arrow-row = { path = "../arrow-rs/arrow-row" }
+arrow-schema = { path = "../arrow-rs/arrow-schema" }
+arrow-select = { path = "../arrow-rs/arrow-select" }
+arrow-string = { path = "../arrow-rs/arrow-string" }
+parquet = { path = "../arrow-rs/parquet" }
-- 
2.50.1
